In [1]:
### let's get some of the necessary libraries in here

import sys
import os
import torch
from torch import nn
import torch.nn.functional as F
import torchvision.transforms as transforms  
import torchvision
import torch.nn as nn
import os
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import time

import cv2
import pandas as pd
import torchvision.transforms as transform 
from torchvision.transforms import ToTensor, Normalize, Resize
from torch.utils.data import DataLoader
from torch.utils.data.sampler import SubsetRandomSampler
from torch.autograd import Variable
from torchvision.datasets import ImageFolder
from torch.utils.data import random_split

In [2]:
dataTrain = "/Users/John/Documents/allProjects/intelData/seg_train/seg_train" 
dataTest = "/Users/John/Documents/allProjects/intelData/seg_test/seg_test"
dataPred = "/Users/John/Documents/allProjects/intelData/seg_pred/seg_pred"

## the folders in the training data give us an easy list of labels
labels = os.listdir(dataTrain)

In [3]:
labels

['forest', 'buildings', 'glacier', 'street', 'mountain', 'sea']

### Normalizing image tensors
This process is important because it helps minimize the impact of image brightness and contrast amongst all images. Unless the images are taken school yearbook style - same location, lighting, camera, etc. - then there will inevitably be some differences among the images.

In [4]:
trainTemp = ImageFolder(dataTrain, transform = transforms.Compose([
    transforms.Resize(64),
    transforms.RandomCrop(64),
    transforms.ToTensor(),
]))
trainDL = DataLoader(trainTemp, 64, shuffle=True)

In [5]:
## do this to extract a single image
for (image, label) in list(enumerate(trainDL))[:1]:
    print(image)

0
0


Our dimensions are [64, 3, 64, 64] so we only need the means and sds for positions 0, 2, 3. We can discared 1 via the dim argument

In [6]:
label[0].size()

torch.Size([64, 3, 64, 64])

In [7]:
def calculateMeanSD(DL):
    """
    This function will calculate the mean and sum of squared mean for the data in a
    DataLoader. I adjusted it specifically to this dataset via the dim argument, skipping
    index 1 because that was not the data required. This function may require other adjustments
    for other datasets.
    """

    ## initialize the three variables as zero
    runningSum, sumSquared, batches = 0,0,0
    
    ## extract the data from the DataLoader and calculate the sums and sum squared
    for data, label in DL:
        runningSum += torch.mean(data, dim = ([0,2,3]))
        sumSquared += torch.mean(data**2, dim = ([0,2,3]))
        batches += 1

    ## simple calcs of     
    mean = runningSum/batches
    std = (sumSquared/batches - mean**2)**0.5
    return mean,std

In [8]:
mean, sd  = calculateMeanSD(trainDL)

Our means and standard deviations are below. This will inform the normalize step of any transformations within the data.

### Augmenting training data
The key here is to add in some randomness so that the CNN detects changes in the image. CNNs are shift invariant, meaning they will detect key details regardless of the position, yet they would struggle if I flipped, cropped, stretched, etc. the images. As such, we'll add in some of that randomness to help the CNN perform better on images with unique traits.

To do this I, resize, add a random crop, a random color jitter, and a random horizontal flip. This is quite a bit of randomness, which should introduce the CNN to a lot of different varieties. I'll likely want to crank the epochs to make sure the CNN "sees" everything.

In [9]:
## we'll do some augmentation on the training data
trainTransform = transform.Compose([
    transform.Resize((64,64)),
    transform.RandomCrop((64,64)),
    transforms.ColorJitter(0.3,0.4,0.4,0.2),
    transform.RandomHorizontalFlip(), ## default is p = 0.5
    transform.ToTensor(),
    transform.Normalize((mean[0],mean[1],mean[2]), (sd[0], sd[1], sd[2]))
])

We do not augment test data because we are evaluating the model's ability to correctly identify the images as opposed to preparing it to identify key attributes anywhere.

In [10]:
testTransform = transform.Compose([
    transform.Resize((64,64)),
    transform.ToTensor(),
    transform.Normalize((mean[0],mean[1],mean[2]), (sd[0], sd[1], sd[2]))
])

In [11]:
train = ImageFolder(dataTrain, transform = trainTransform)
test = ImageFolder(dataTest, transform = testTransform)

## Modeling

In [12]:
seedA = 33
torch.manual_seed(seedA);

In [13]:
## set the sizes and train test split the data
train_size = int(0.8*len(train))
val_size = int(len(train) - train_size)
print(train_size, val_size)

trainData, valData = random_split(train, [train_size, val_size])

11227 2807
11227 2807


Set parameters and the device. We'd like to use my gpu if it is available. If it is not, we can use the CPU.

In [14]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
batch_size = 64
classes = len(labels)
learning_rate = 0.001
epochs = 50

In [15]:
## define the dataloaders
train_dl = DataLoader(trainData, batch_size, shuffle=True, num_workers=2, pin_memory=True)
valid_dl = DataLoader(valData, batch_size*2, num_workers=2, pin_memory=True)
test_dl = DataLoader(test, batch_size*2, num_workers=2, pin_memory=True)

### Definine the Neural Network

I'm going to do one bespoke CNN and one pretrained model for comparison purposes. To build a bespoke NN with pytorch we need to define the class with nn.Module. 

In [16]:
class BespokeCNN(nn.Module):
    """
    This is an extremely basic CNN, I doubt it will perform as well as the
    pretrained models.
    """
	## Decide upon some layers within the network
    def __init__(self, classes):
        super(BespokeCNN, self).__init__()
        self.conv1 = nn.Conv2d(in_channels=3, out_channels=32, kernel_size=3)
        self.max_pool1 = nn.MaxPool2d(kernel_size = 2, stride = 2)
        self.conv2 = nn.Conv2d(in_channels=32, out_channels=64, kernel_size=3)
        self.max_pool2 = nn.MaxPool2d(kernel_size = 2, stride = 2)

        ## fully connected layers with Softmax as the activation function
        ## Using softmax because this is a multi-classification problem
        self.fc1 = nn.Linear(12544, 128)
        self.fc2 = nn.Linear(128, classes)
        self.softmax = nn.Softmax(dim = -1)
        
    
    ## defines how the data moves across the layers
    def forward(self,x):
        forw = self.conv1(x)
        forw = self.max_pool1(forw)
        
        forw = self.conv2(forw)
        forw = self.max_pool2(forw)
                
        forw = forw.reshape(forw.size(0), -1)
        
        forw = self.fc1(forw)
        forw = self.fc2(forw)
        forw = self.softmax(forw)
        

        return forw

In [28]:
## define the model from the class we built
model = BespokeCNN(classes)

In [22]:
def trainModel(model, epochs, train_dl, valid_dl, learning_rate):
    """
    This function will train the provided model printing the accuracy for the validation data.
    It takes the model, number of epochs, and two data loaders and then runs the model through the 
    training process. If we want to change the loss function or optimizer, we can in the code below
    """
    criterion = nn.CrossEntropyLoss()
    optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, weight_decay = 0.001, momentum = 0.9)  
    #total_step = len(train_dl)


    for epoch in range(epochs):
        start = time.time()
        #Load in the data in batches using the train_loader object
        for i, (images, labels) in enumerate(train_dl):  
            # Move tensors to the configured device
            images = images.to(device)
            labels = labels.to(device)
            
            # Forward pass
            outputs = model(images)
            loss = criterion(outputs, labels)
            
            # Backward and optimize
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

        ## let's monitor the prediction accuracy on the validation data for each epoch
        ## this may indicate a good stopping point for the particular neural network
        with torch.no_grad():
            correct = 0
            total = 0
            for images, labels in valid_dl:
                images = images.to(device)
                labels = labels.to(device)
                outputs = model(images)
                _, predicted = torch.max(outputs.data, 1)
                total += labels.size(0)
                correct += (predicted == labels).sum().item()
            print('Accuracy on {} validation images: {:.4f} %'.format(val_size, 100 * correct / total))

        end = time.time()
        print('Epoch [{}/{}], Loss: {:.4f}, Time in Seconds: {:.4f}'.format(epoch+1, epochs, loss.item(), end - start))

    return model

In [24]:
modelOut = trainModel(model, 25, train_dl, valid_dl, learning_rate) 

Accuracy on 2807 validation images: 52.7253 %
Epoch [1/25], Loss: 1.5926, Time in Seconds: 24.4088
Accuracy on 2807 validation images: 52.7253 %
Epoch [1/25], Loss: 1.5926, Time in Seconds: 24.4088


Accuracy on 2807 validation images: 54.9341 %
Epoch [2/25], Loss: 1.5307, Time in Seconds: 23.9522
Accuracy on 2807 validation images: 54.9341 %
Epoch [2/25], Loss: 1.5307, Time in Seconds: 23.9522


Accuracy on 2807 validation images: 54.9341 %
Epoch [3/25], Loss: 1.5617, Time in Seconds: 23.7382
Accuracy on 2807 validation images: 54.9341 %
Epoch [3/25], Loss: 1.5617, Time in Seconds: 23.7382


Accuracy on 2807 validation images: 55.8603 %
Epoch [4/25], Loss: 1.4687, Time in Seconds: 23.8442
Accuracy on 2807 validation images: 55.8603 %
Epoch [4/25], Loss: 1.4687, Time in Seconds: 23.8442


Accuracy on 2807 validation images: 56.5372 %
Epoch [5/25], Loss: 1.3823, Time in Seconds: 23.3743
Accuracy on 2807 validation images: 56.5372 %
Epoch [5/25], Loss: 1.3823, Time in Seconds: 23.3743


Accuracy on 2807 validation images: 56.1454 %
Epoch [6/25], Loss: 1.3027, Time in Seconds: 23.8263
Accuracy on 2807 validation images: 56.1454 %
Epoch [6/25], Loss: 1.3027, Time in Seconds: 23.8263


Accuracy on 2807 validation images: 57.3922 %
Epoch [7/25], Loss: 1.4306, Time in Seconds: 23.6338
Accuracy on 2807 validation images: 57.3922 %
Epoch [7/25], Loss: 1.4306, Time in Seconds: 23.6338


Accuracy on 2807 validation images: 59.1022 %
Epoch [8/25], Loss: 1.5030, Time in Seconds: 23.8514
Accuracy on 2807 validation images: 59.1022 %
Epoch [8/25], Loss: 1.5030, Time in Seconds: 23.8514


Accuracy on 2807 validation images: 58.4966 %
Epoch [9/25], Loss: 1.2865, Time in Seconds: 23.8525
Accuracy on 2807 validation images: 58.4966 %
Epoch [9/25], Loss: 1.2865, Time in Seconds: 23.8525


Accuracy on 2807 validation images: 58.3185 %
Epoch [10/25], Loss: 1.4515, Time in Seconds: 23.5675
Accuracy on 2807 validation images: 58.3185 %
Epoch [10/25], Loss: 1.4515, Time in Seconds: 23.5675


Accuracy on 2807 validation images: 60.5629 %
Epoch [11/25], Loss: 1.2663, Time in Seconds: 23.7524
Accuracy on 2807 validation images: 60.5629 %
Epoch [11/25], Loss: 1.2663, Time in Seconds: 23.7524


Accuracy on 2807 validation images: 59.6366 %
Epoch [12/25], Loss: 1.4705, Time in Seconds: 23.5645
Accuracy on 2807 validation images: 59.6366 %
Epoch [12/25], Loss: 1.4705, Time in Seconds: 23.5645


Accuracy on 2807 validation images: 60.7054 %
Epoch [13/25], Loss: 1.4525, Time in Seconds: 23.9887
Accuracy on 2807 validation images: 60.7054 %
Epoch [13/25], Loss: 1.4525, Time in Seconds: 23.9887


Accuracy on 2807 validation images: 62.0235 %
Epoch [14/25], Loss: 1.4016, Time in Seconds: 23.7018
Accuracy on 2807 validation images: 62.0235 %
Epoch [14/25], Loss: 1.4016, Time in Seconds: 23.7018


Accuracy on 2807 validation images: 60.4560 %
Epoch [15/25], Loss: 1.4029, Time in Seconds: 23.4981
Accuracy on 2807 validation images: 60.4560 %
Epoch [15/25], Loss: 1.4029, Time in Seconds: 23.4981


Accuracy on 2807 validation images: 62.5579 %
Epoch [16/25], Loss: 1.4704, Time in Seconds: 23.9585
Accuracy on 2807 validation images: 62.5579 %
Epoch [16/25], Loss: 1.4704, Time in Seconds: 23.9585


Accuracy on 2807 validation images: 63.2348 %
Epoch [17/25], Loss: 1.3993, Time in Seconds: 23.6773
Accuracy on 2807 validation images: 63.2348 %
Epoch [17/25], Loss: 1.3993, Time in Seconds: 23.6773


Accuracy on 2807 validation images: 62.7360 %
Epoch [18/25], Loss: 1.4288, Time in Seconds: 23.7531
Accuracy on 2807 validation images: 62.7360 %
Epoch [18/25], Loss: 1.4288, Time in Seconds: 23.7531


Accuracy on 2807 validation images: 64.8735 %
Epoch [19/25], Loss: 1.3159, Time in Seconds: 23.6075
Accuracy on 2807 validation images: 64.8735 %
Epoch [19/25], Loss: 1.3159, Time in Seconds: 23.6075


Accuracy on 2807 validation images: 63.3060 %
Epoch [20/25], Loss: 1.3714, Time in Seconds: 23.4025
Accuracy on 2807 validation images: 63.3060 %
Epoch [20/25], Loss: 1.3714, Time in Seconds: 23.4025


Accuracy on 2807 validation images: 64.0542 %
Epoch [21/25], Loss: 1.4479, Time in Seconds: 24.8049
Accuracy on 2807 validation images: 64.0542 %
Epoch [21/25], Loss: 1.4479, Time in Seconds: 24.8049


Accuracy on 2807 validation images: 63.1635 %
Epoch [22/25], Loss: 1.3894, Time in Seconds: 24.2381
Accuracy on 2807 validation images: 63.1635 %
Epoch [22/25], Loss: 1.3894, Time in Seconds: 24.2381


Accuracy on 2807 validation images: 64.7310 %
Epoch [23/25], Loss: 1.4283, Time in Seconds: 23.6714
Accuracy on 2807 validation images: 64.7310 %
Epoch [23/25], Loss: 1.4283, Time in Seconds: 23.6714


Accuracy on 2807 validation images: 63.4129 %
Epoch [24/25], Loss: 1.3788, Time in Seconds: 24.1553
Accuracy on 2807 validation images: 63.4129 %
Epoch [24/25], Loss: 1.3788, Time in Seconds: 24.1553


Accuracy on 2807 validation images: 65.3723 %
Epoch [25/25], Loss: 1.3588, Time in Seconds: 24.7434
Accuracy on 2807 validation images: 65.3723 %
Epoch [25/25], Loss: 1.3588, Time in Seconds: 24.7434


At 25 epochs the loss function is headed downwards and the prediction accuracy is increasing. This suggests more training cycles could be helpful. I'm going to re-run with 50 to see what I get.

In [29]:
modelOut = trainModel(model, 50, train_dl, valid_dl, learning_rate) 

Accuracy on 2807 validation images: 42.7146 %
Epoch [1/50], Loss: 1.6391, Time in Seconds: 25.3216
Accuracy on 2807 validation images: 42.7146 %
Epoch [1/50], Loss: 1.6391, Time in Seconds: 25.3216


Accuracy on 2807 validation images: 48.8422 %
Epoch [2/50], Loss: 1.5251, Time in Seconds: 24.1999
Accuracy on 2807 validation images: 48.8422 %
Epoch [2/50], Loss: 1.5251, Time in Seconds: 24.1999


Accuracy on 2807 validation images: 52.6185 %
Epoch [3/50], Loss: 1.3587, Time in Seconds: 23.9948
Accuracy on 2807 validation images: 52.6185 %
Epoch [3/50], Loss: 1.3587, Time in Seconds: 23.9948


Accuracy on 2807 validation images: 52.9035 %
Epoch [4/50], Loss: 1.6053, Time in Seconds: 24.3778
Accuracy on 2807 validation images: 52.9035 %
Epoch [4/50], Loss: 1.6053, Time in Seconds: 24.3778


Accuracy on 2807 validation images: 54.0791 %
Epoch [5/50], Loss: 1.5078, Time in Seconds: 23.8178
Accuracy on 2807 validation images: 54.0791 %
Epoch [5/50], Loss: 1.5078, Time in Seconds: 23.8178


Accuracy on 2807 validation images: 54.9341 %
Epoch [6/50], Loss: 1.5962, Time in Seconds: 23.9399
Accuracy on 2807 validation images: 54.9341 %
Epoch [6/50], Loss: 1.5962, Time in Seconds: 23.9399


Accuracy on 2807 validation images: 54.6847 %
Epoch [7/50], Loss: 1.5428, Time in Seconds: 23.8549
Accuracy on 2807 validation images: 54.6847 %
Epoch [7/50], Loss: 1.5428, Time in Seconds: 23.8549


Accuracy on 2807 validation images: 56.6797 %
Epoch [8/50], Loss: 1.5144, Time in Seconds: 23.5247
Accuracy on 2807 validation images: 56.6797 %
Epoch [8/50], Loss: 1.5144, Time in Seconds: 23.5247


Accuracy on 2807 validation images: 56.1454 %
Epoch [9/50], Loss: 1.5776, Time in Seconds: 23.7885
Accuracy on 2807 validation images: 56.1454 %
Epoch [9/50], Loss: 1.5776, Time in Seconds: 23.7885


Accuracy on 2807 validation images: 56.7510 %
Epoch [10/50], Loss: 1.3387, Time in Seconds: 24.4587
Accuracy on 2807 validation images: 56.7510 %
Epoch [10/50], Loss: 1.3387, Time in Seconds: 24.4587


Accuracy on 2807 validation images: 57.8197 %
Epoch [11/50], Loss: 1.3349, Time in Seconds: 24.1277
Accuracy on 2807 validation images: 57.8197 %
Epoch [11/50], Loss: 1.3349, Time in Seconds: 24.1277


Accuracy on 2807 validation images: 58.9597 %
Epoch [12/50], Loss: 1.4548, Time in Seconds: 25.3639
Accuracy on 2807 validation images: 58.9597 %
Epoch [12/50], Loss: 1.4548, Time in Seconds: 25.3639


Accuracy on 2807 validation images: 59.5654 %
Epoch [13/50], Loss: 1.3941, Time in Seconds: 24.7617
Accuracy on 2807 validation images: 59.5654 %
Epoch [13/50], Loss: 1.3941, Time in Seconds: 24.7617


Accuracy on 2807 validation images: 59.5297 %
Epoch [14/50], Loss: 1.3212, Time in Seconds: 24.3005
Accuracy on 2807 validation images: 59.5297 %
Epoch [14/50], Loss: 1.3212, Time in Seconds: 24.3005


Accuracy on 2807 validation images: 60.7054 %
Epoch [15/50], Loss: 1.5078, Time in Seconds: 24.0327
Accuracy on 2807 validation images: 60.7054 %
Epoch [15/50], Loss: 1.5078, Time in Seconds: 24.0327


Accuracy on 2807 validation images: 59.4941 %
Epoch [16/50], Loss: 1.3955, Time in Seconds: 24.7815
Accuracy on 2807 validation images: 59.4941 %
Epoch [16/50], Loss: 1.3955, Time in Seconds: 24.7815


Accuracy on 2807 validation images: 58.7460 %
Epoch [17/50], Loss: 1.5382, Time in Seconds: 23.9968
Accuracy on 2807 validation images: 58.7460 %
Epoch [17/50], Loss: 1.5382, Time in Seconds: 23.9968


Accuracy on 2807 validation images: 60.5273 %
Epoch [18/50], Loss: 1.5624, Time in Seconds: 24.0338
Accuracy on 2807 validation images: 60.5273 %
Epoch [18/50], Loss: 1.5624, Time in Seconds: 24.0338


Accuracy on 2807 validation images: 60.9191 %
Epoch [19/50], Loss: 1.5547, Time in Seconds: 24.2441
Accuracy on 2807 validation images: 60.9191 %
Epoch [19/50], Loss: 1.5547, Time in Seconds: 24.2441


Accuracy on 2807 validation images: 60.5985 %
Epoch [20/50], Loss: 1.2432, Time in Seconds: 24.1416
Accuracy on 2807 validation images: 60.5985 %
Epoch [20/50], Loss: 1.2432, Time in Seconds: 24.1416


Accuracy on 2807 validation images: 62.2373 %
Epoch [21/50], Loss: 1.4530, Time in Seconds: 23.7670
Accuracy on 2807 validation images: 62.2373 %
Epoch [21/50], Loss: 1.4530, Time in Seconds: 23.7670


Accuracy on 2807 validation images: 62.4154 %
Epoch [22/50], Loss: 1.5444, Time in Seconds: 24.1739
Accuracy on 2807 validation images: 62.4154 %
Epoch [22/50], Loss: 1.5444, Time in Seconds: 24.1739


Accuracy on 2807 validation images: 62.8785 %
Epoch [23/50], Loss: 1.3406, Time in Seconds: 24.3020
Accuracy on 2807 validation images: 62.8785 %
Epoch [23/50], Loss: 1.3406, Time in Seconds: 24.3020


Accuracy on 2807 validation images: 62.2373 %
Epoch [24/50], Loss: 1.2806, Time in Seconds: 24.5073
Accuracy on 2807 validation images: 62.2373 %
Epoch [24/50], Loss: 1.2806, Time in Seconds: 24.5073


Accuracy on 2807 validation images: 63.3416 %
Epoch [25/50], Loss: 1.4416, Time in Seconds: 25.1522
Accuracy on 2807 validation images: 63.3416 %
Epoch [25/50], Loss: 1.4416, Time in Seconds: 25.1522


Accuracy on 2807 validation images: 62.0948 %
Epoch [26/50], Loss: 1.4307, Time in Seconds: 24.6753
Accuracy on 2807 validation images: 62.0948 %
Epoch [26/50], Loss: 1.4307, Time in Seconds: 24.6753


Accuracy on 2807 validation images: 64.5173 %
Epoch [27/50], Loss: 1.5421, Time in Seconds: 23.6883
Accuracy on 2807 validation images: 64.5173 %
Epoch [27/50], Loss: 1.5421, Time in Seconds: 23.6883


Accuracy on 2807 validation images: 63.5910 %
Epoch [28/50], Loss: 1.2988, Time in Seconds: 25.1394
Accuracy on 2807 validation images: 63.5910 %
Epoch [28/50], Loss: 1.2988, Time in Seconds: 25.1394


Accuracy on 2807 validation images: 64.4104 %
Epoch [29/50], Loss: 1.2871, Time in Seconds: 23.8512
Accuracy on 2807 validation images: 64.4104 %
Epoch [29/50], Loss: 1.2871, Time in Seconds: 23.8512


Accuracy on 2807 validation images: 64.8379 %
Epoch [30/50], Loss: 1.5197, Time in Seconds: 23.3971
Accuracy on 2807 validation images: 64.8379 %
Epoch [30/50], Loss: 1.5197, Time in Seconds: 23.3971


Accuracy on 2807 validation images: 63.8760 %
Epoch [31/50], Loss: 1.3158, Time in Seconds: 24.2176
Accuracy on 2807 validation images: 63.8760 %
Epoch [31/50], Loss: 1.3158, Time in Seconds: 24.2176


Accuracy on 2807 validation images: 65.1942 %
Epoch [32/50], Loss: 1.4810, Time in Seconds: 24.2327
Accuracy on 2807 validation images: 65.1942 %
Epoch [32/50], Loss: 1.4810, Time in Seconds: 24.2327


Accuracy on 2807 validation images: 64.1254 %
Epoch [33/50], Loss: 1.3319, Time in Seconds: 24.2471
Accuracy on 2807 validation images: 64.1254 %
Epoch [33/50], Loss: 1.3319, Time in Seconds: 24.2471


Accuracy on 2807 validation images: 65.6929 %
Epoch [34/50], Loss: 1.2282, Time in Seconds: 24.0863
Accuracy on 2807 validation images: 65.6929 %
Epoch [34/50], Loss: 1.2282, Time in Seconds: 24.0863


Accuracy on 2807 validation images: 65.2298 %
Epoch [35/50], Loss: 1.2275, Time in Seconds: 23.8344
Accuracy on 2807 validation images: 65.2298 %
Epoch [35/50], Loss: 1.2275, Time in Seconds: 23.8344


Accuracy on 2807 validation images: 65.2654 %
Epoch [36/50], Loss: 1.3097, Time in Seconds: 24.2290
Accuracy on 2807 validation images: 65.2654 %
Epoch [36/50], Loss: 1.3097, Time in Seconds: 24.2290


Accuracy on 2807 validation images: 65.3367 %
Epoch [37/50], Loss: 1.2854, Time in Seconds: 24.1789
Accuracy on 2807 validation images: 65.3367 %
Epoch [37/50], Loss: 1.2854, Time in Seconds: 24.1789


Accuracy on 2807 validation images: 65.6217 %
Epoch [38/50], Loss: 1.4903, Time in Seconds: 23.7771
Accuracy on 2807 validation images: 65.6217 %
Epoch [38/50], Loss: 1.4903, Time in Seconds: 23.7771


Accuracy on 2807 validation images: 65.4079 %
Epoch [39/50], Loss: 1.5298, Time in Seconds: 23.7807
Accuracy on 2807 validation images: 65.4079 %
Epoch [39/50], Loss: 1.5298, Time in Seconds: 23.7807


Accuracy on 2807 validation images: 66.6548 %
Epoch [40/50], Loss: 1.2338, Time in Seconds: 24.2001
Accuracy on 2807 validation images: 66.6548 %
Epoch [40/50], Loss: 1.2338, Time in Seconds: 24.2001


Accuracy on 2807 validation images: 67.1535 %
Epoch [41/50], Loss: 1.4182, Time in Seconds: 23.9306
Accuracy on 2807 validation images: 67.1535 %
Epoch [41/50], Loss: 1.4182, Time in Seconds: 23.9306


Accuracy on 2807 validation images: 66.5123 %
Epoch [42/50], Loss: 1.3421, Time in Seconds: 24.5953
Accuracy on 2807 validation images: 66.5123 %
Epoch [42/50], Loss: 1.3421, Time in Seconds: 24.5953


Accuracy on 2807 validation images: 67.6523 %
Epoch [43/50], Loss: 1.3574, Time in Seconds: 23.7444
Accuracy on 2807 validation images: 67.6523 %
Epoch [43/50], Loss: 1.3574, Time in Seconds: 23.7444


Accuracy on 2807 validation images: 67.6167 %
Epoch [44/50], Loss: 1.4547, Time in Seconds: 23.8040
Accuracy on 2807 validation images: 67.6167 %
Epoch [44/50], Loss: 1.4547, Time in Seconds: 23.8040


Accuracy on 2807 validation images: 66.7260 %
Epoch [45/50], Loss: 1.4221, Time in Seconds: 23.6053
Accuracy on 2807 validation images: 66.7260 %
Epoch [45/50], Loss: 1.4221, Time in Seconds: 23.6053


Accuracy on 2807 validation images: 66.4410 %
Epoch [46/50], Loss: 1.2321, Time in Seconds: 23.5884
Accuracy on 2807 validation images: 66.4410 %
Epoch [46/50], Loss: 1.2321, Time in Seconds: 23.5884


Accuracy on 2807 validation images: 66.6192 %
Epoch [47/50], Loss: 1.2753, Time in Seconds: 23.7463
Accuracy on 2807 validation images: 66.6192 %
Epoch [47/50], Loss: 1.2753, Time in Seconds: 23.7463


Accuracy on 2807 validation images: 67.4385 %
Epoch [48/50], Loss: 1.3611, Time in Seconds: 24.0111
Accuracy on 2807 validation images: 67.4385 %
Epoch [48/50], Loss: 1.3611, Time in Seconds: 24.0111


Accuracy on 2807 validation images: 65.7642 %
Epoch [49/50], Loss: 1.3243, Time in Seconds: 23.8575
Accuracy on 2807 validation images: 65.7642 %
Epoch [49/50], Loss: 1.3243, Time in Seconds: 23.8575


Accuracy on 2807 validation images: 67.2960 %
Epoch [50/50], Loss: 1.4187, Time in Seconds: 24.0643
Accuracy on 2807 validation images: 67.2960 %
Epoch [50/50], Loss: 1.4187, Time in Seconds: 24.0643


The model seems to top out at 67% accuracy. Compared to other models, this is not very good. Models like ResNet9 have gotten a 92% accuracy on the same data. Reviewing why:

- This is not deep. The famously strong models are deeper and use extra bells and whistles in the training process.
- The model is simple. Two convolutional and pooling layers with two fully connected layers before a softmax activation function is probably not going to get the job done. It's hardly convolutional, and some might even say that it isn't.
- Low model capacity. I intentionally set this low because I'm running this on my personal CPU, as such, a model with higher capacity will likely take eons to train due to my lack of a GPU

#### Summary
Ultimately, I would want to implement a well-known model or two and compare the performance. This exercise was meant to show my ability to create, implement, train and evaluate a bespoke neural network. Success! The model is poor, but it is a model built from scratch. If I were chasing performance I would go about this **very differently**. The reader would never see this bespoke model because I gravitate towards well-researched approaches. There is a lot of peer-review literature in this world, I am not smarter than the folks who do this research for their careers.

#### Thank you for making it this far!
